metricbeat/module/vsphere/datastore/datastore.go

```diff
@@ -81,44 +102,137 @@ func (m *MetricSet) Fetch(ctx context.Context, reporter mb.ReporterV2) error {
 
 	defer func() {
 		if err := v.Destroy(ctx); err != nil {
-			m.Logger().Debug(fmt.Errorf("error trying to destroy view from vshphere: %w", err))
+			m.Logger().Debugf("error trying to destroy view from vSphere: %w", err)
 		}
 	}()
 
 	// Retrieve summary property for all datastores
 	var dst []mo.Datastore
-	if err = v.Retrieve(ctx, []string{"Datastore"}, []string{"summary"}, &dst); err != nil {
+	err = v.Retrieve(ctx, []string{"Datastore"}, []string{"summary", "host", "vm", "overallStatus"}, &dst)
+	if err != nil {
 		return fmt.Errorf("error in Retrieve: %w", err)
 	}
 
-	for _, ds := range dst {
-		var usedSpacePercent float64
-		if ds.Summary.Capacity > 0 {
-			usedSpacePercent = float64(ds.Summary.Capacity-ds.Summary.FreeSpace) / float64(ds.Summary.Capacity)
-		}
-		usedSpaceBytes := ds.Summary.Capacity - ds.Summary.FreeSpace
-
-		event := mapstr.M{
-			"name":   ds.Summary.Name,
-			"fstype": ds.Summary.Type,
-			"capacity": mapstr.M{
-				"total": mapstr.M{
-					"bytes": ds.Summary.Capacity,
-				},
-				"free": mapstr.M{
-					"bytes": ds.Summary.FreeSpace,
-				},
-				"used": mapstr.M{
-					"bytes": usedSpaceBytes,
-					"pct":   usedSpacePercent,
-				},
-			},
+	// Create a performance manager
+	perfManager := performance.NewManager(c)
+
+	// Retrieve all available metrics
+	metrics, err := perfManager.CounterInfoByName(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve metrics: %w", err)
+	}
+
+	// Filter for required metrics
+	var metricIds []types.PerfMetricId
+	for metricName := range metricSet {
+		if metric, ok := metrics[metricName]; ok {
+			metricIds = append(metricIds, types.PerfMetricId{CounterId: metric.Key})
+		} else {
+			m.Logger().Warnf("Metric %s not found", metricName)
```

Comment:
Used the map directly
```suggestion
	for metricName := range map[string]struct{}{
		"datastore.read.average":              {},
		"datastore.write.average":             {},
		"datastore.datastoreIops.average":     {},
		"datastore.totalReadLatency.average":  {},
		"datastore.totalWriteLatency.average": {},
	} {
		if metric, ok := metrics[metricName]; ok {
			metricIds = append(metricIds, types.PerfMetricId{CounterId: metric.Key})
			continue
		}
		m.Logger().Warnf("Metric %s not found", metricName)
	}
```